In [ ]:
from bs4 import BeautifulSoup
import requests

response = requests.get("https://www.udacity.com/courses/all")
html_soup = BeautifulSoup(response.content, 'html.parser')

#find all the URLs (items in the html where href exists)
url = html_soup.find_all(href=True)

#Iterate through the urls, and only return the ones that contain course and --
links = []
for u in url:
    if (u['href'].find('course') != -1):
        if (u['href'].find('--') != -1):
            if (u['href'].find('nanodegree') == -1 and u['href'].find('--nd') == -1):
                links.append(u['href'])
            
#remove duplicates by converting the list of links to a set, then back to a list
links = list(set(links)) 
links

['/course/data-analysis-with-r--ud651',
 '/course/machine-learning-for-trading--ud501',
 '/course/gt-refresher-advanced-os--ud098',
 '/course/intro-to-psychology--ps001',
 '/course/auto-layout--ud1026',
 '/course/how-to-install-android-studio--ud808',
 '/course/intro-to-data-science--ud359',
 '/course/objective-c-for-swift-developers--ud1009',
 '/course/java-programming-basics--ud282',
 '/course/intro-to-backend--ud171',
 '/course/android-performance--ud825',
 '/course/mobile-design-and-usability-for-android--ud358',
 '/course/app-monetization--ud518',
 '/course/classification-models--ud978',
 '/course/networking-for-web-developers--ud256',
 '/course/artificial-intelligence-for-robotics--cs373',
 '/course/engagement-monetization-mobile-games--ud407',
 '/course/build-native-mobile-apps-with-flutter--ud905',
 '/course/computer-networking--ud436',
 '/course/web-accessibility--ud891',
 '/course/firebase-analytics-ios--ud353',
 '/course/intel-edge-AI-fundamentals-with-openvino--ud132',
 '/c

In [ ]:
len(links)

186

186 courses : 

Each course URL will be visited in the following code to extract the information needed. The relevent HTML tags and class names were extracted by visiting the URL, right clicking on the element of interest and selecting Inspect to view the related HTML. This could then be used with the BeautifulSoup find and find_all functions to get the information.

In [ ]:
import time
from random import randint

#initilise the lists to store the extracted info
course_names = []
course_urls = []
course_nanodegrees = []
course_costs = []

# Preparing the monitoring of the loop
start_time = time.time()
requests_count = 0

#iterate through each link to extract info about the nano degree it belongs to and the cost
for i in range(len(links)):
    #need to make requests at random times in order to not be blocked by website
    time.sleep(randint(2,6)) 
    url = "https://www.udacity.com" + links[i]
    response = requests.get(url)
    content = response.content
    html_soup = BeautifulSoup(content, 'html.parser')
    
    # Monitor the requests
    requests_count += 1
    elapsed_time = time.time() - start_time
    print('Request:{}; Frequency: {} requests/s'.format(requests_count, requests_count/elapsed_time))
    #clear_output(wait = True)
    
    # Break the loop if the number of requests is greater than expected
    if requests_count > 250:
        warn('Number of requests was greater than expected.')  
        break 
    
    try:
        nanodegree = html_soup.find('h3',class_="h1 hero__nanodegree--title").text
    except:
        nanodegree = 'None'
    try:
        course_name = html_soup.find('h1',class_='hero__course--title').text
    except:
        course_name = 'None'
    try:
        cost = html_soup.find('h6',class_='hero__course--type').text
    except:
        cost = 'None'
    course_names.append(course_name)
    course_urls.append(url)
    course_nanodegrees.append(nanodegree)
    course_costs.append(cost)


Request:1; Frequency: 0.30723007044317396 requests/s
Request:2; Frequency: 0.26649758787603917 requests/s
Request:3; Frequency: 0.2778304090832433 requests/s
Request:4; Frequency: 0.28441414771033285 requests/s
Request:5; Frequency: 0.30612204383735303 requests/s
Request:6; Frequency: 0.3210213893891675 requests/s
Request:7; Frequency: 0.2804222609796274 requests/s
Request:8; Frequency: 0.27365660956369336 requests/s
Request:9; Frequency: 0.2762723613554465 requests/s
Request:10; Frequency: 0.2711005960775285 requests/s
Request:11; Frequency: 0.2532987472980598 requests/s
Request:12; Frequency: 0.24616476916264565 requests/s
Request:13; Frequency: 0.25470122846014953 requests/s
Request:14; Frequency: 0.2526767058029134 requests/s
Request:15; Frequency: 0.2421179306250676 requests/s
Request:16; Frequency: 0.24123349609279235 requests/s
Request:17; Frequency: 0.2340058565223791 requests/s
Request:18; Frequency: 0.2298633545083746 requests/s
Request:19; Frequency: 0.23188989315948383 requ

In [ ]:
#Convert the lists to a dataframe
import pandas as pd
courses_df = pd.DataFrame({'Name': course_names,
                          'URL': course_urls,
                          'Nanodegree': course_nanodegrees,
                          'Cost':course_costs})
courses_df = courses_df.sort_values('Nanodegree')
print(courses_df.info())
courses_df

<class 'pandas.core.frame.DataFrame'>
Int64Index: 186 entries, 171 to 143
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Name        186 non-null    object
 1   URL         186 non-null    object
 2   Nanodegree  186 non-null    object
 3   Cost        186 non-null    object
dtypes: object(4)
memory usage: 7.3+ KB
None


Name  ...         Cost
171          Android Basics: Multiscreen Apps  ...  Free Course
145                Android Basics: User Input  ...  Free Course
115                Android Basics: Networking  ...  Free Course
53               Android Basics: Data Storage  ...  Free Course
119       Developing Android Apps with Kotlin  ...  Free Course
..                                        ...  ...          ...
136                  Version Control with Git  ...  Free Course
3                         Intro to Psychology  ...  Free Course
131  Cyber-Physical Systems Design & Analysis  ...  Free Course
49                       Swift for Developers  ...  Free Course
143       Introduction to Graduate Algorithms  ...  Free Course

[186 rows x 4 columns]

In [ ]:
#Save to csv file
courses_df.to_csv('Udacity_Courses.csv')

To see which courses belonged to the Data Analyst degree, the courses data frame was filtered and a new data frame was created

In [ ]:
data_analyst = courses_df[courses_df['Nanodegree'] == 'Data Analyst']
data_analyst

Name  ...         Cost
148       Human-Computer Interaction  ...  Free Course
141  Data Analysis and Visualization  ...  Free Course
169      Software Analysis & Testing  ...  Free Course
23   Intro to Descriptive Statistics  ...  Free Course

[4 rows x 4 columns]

In [ ]:
from google.colab import files
files.download('Udacity_Courses.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>